Added assertions to protect correct block sizes in CUDA reductions

DavidPoliakoff opened the pull request:
Small code change, added static asserts testing 

1) That the block size is a power of two
2) That the block size is between 32 and 2048. 2048 is chosen randomly, feel free to propose changes

Reviews from either of @jonesholger or @davidbeckingsale ? You'll need to build CUDA, do a correctly sized reduction, and an incorrectly sized reduction to test. To remove the WIP tag we need to settle on a max block size

Note: code contributed by @trws

rhornung67 commented:
@DavidPoliakoff I approve.  Do we need to add such things in any other places in the code?

DavidPoliakoff commented:
Almost certainly. Specifically, we probably need these exact assertions in other spots (will do that myself). More generally (as discussed in the meeting) static_asserts make for a much more understandable system. @trws actually has some really nifty ones in his RAJA overhaul IIRC. Anyway, I'll see where else these need to be if you're comfortable. Also, I'll make a RAJA_CUDA_MAX_BLOCKS_SIZE in the config file instead of 2048

DavidPoliakoff commented:
Actually, I'll encode it as 2048 in the raja_cuda.hxx file, I don't want to put CUDA-specific info in a non-CUDA file

trws commented:
Yeah, it would be good to work them in as we go so users get better compile-time error messages. It kinda has to be an as we go thing though, because there aren't always well-defined invariants to check just yet.  

One of the ones @DavidPoliakoff is referring to is my iterator rework actually checks to make sure that the object passed in as a segment has the required interface, and prints an error with a note about the required type if not.  The forallN stuff ensures that the number of parameters matches the number of arguments and that the lambda can be converted to a type that accepts that number of arguments, along with checking to make sure that the thing passed as the last argument actually is a callable in the first place.  These are useful, but things like the INDEXSET_T template I have no idea how to do a check for, because there's no standardized definition for what that should take (though for a while I had one in there that checked that the object passed had to be a descendent of IndexSet, that kinda defeated the purpose of that function because of the IndexSet& overload).

DavidPoliakoff commented:
This is a little off-topic, but I'll just say that a lot of the reason I use Clang is that the error messages just make sense, so I'd seriously think about this kind of stuff. The moment we send a user some "no valid version of function detected for foo(), here's 65000 possibilities" rather than "hey, you should have used a power of two" we've lost a user.

Anyway, I implemented the functionality, commented it, and moved it to the bottom of the class so it doesn't distract from the code itself

jonesholger commented:
I think this will also play into the work we're putting into the 3d-generic reductions and the kernel launch configurations we anticipate. We're building some additional unit tests to exercise this. David B. is putting together some tests, but will probably need to wait for Adam to help flesh out all of the the forall_N variants so we get the proper coverage

@DavidPoliakoff thanks for doing this.  I'll merge this into my 3d-generic-reductions and do some tests tomorrow

trws commented:
@rhornung67, I notice you say you approve, is that plus the green test light sufficient for a merge?

rhornung67 commented:
@trws At this point, that works for me. Until we have automated testing set up (next now that the build stuff is squared away), we will have to be vigilant about running some tests manually before committing/merging.

trws commented:
Fair enough.  I'll merge this now, and just put up a PR so that we actually have at least *some* automated testing while we wait for the cleaner testing PR to drop.
